### PR TITLE
Parallel Sobol

### PR DESCRIPTION
**Pull request by bernardoct**
Shared memory parallel version of Sobol analysis. Tested with Ishigami function.

### CODE REVIEW

**Comment by willu47** [comment]
Hi @bernardoct.  Thanks a lot for the pull request!  Great to see that you've been using the library.  I know the @jdherman is having a detailed look at your work so I won't comment until he's had a look through.  One quick general niggle - generally it's best to create a new branch (e.g. called develop) rather than working off your master.  This avoids confusion when merging in pull requests as there are now two branches called master.

__UnitTests__

If you'd like to earn extra kudos :+1: , writing some unit tests for the functions `first_order`, `second_order`, `total_order` and `sobol_indexes_calculation` would be amazing!  You can see that the gaps in our test coverage [here](https://coveralls.io/builds/3419077/source?filename=SALib%2Fanalyze%2Fsobol.py)  :(.

**Comment by bernardoct** [comment]
Hi @willu47. I'm more than glad to be of help! I'm sorry if I made a mess, but I e-mailed @jdherman yesterday and asked him how he would prefer me to do it and said it was customary for y'all to fork SALib to your own github account and then send a pull request. 

I talked with @jdherman yesterday on Skype (we used to be office mates at Cornell) and as soon as we figure his comments out I'll come up with the tests for the functions you mentioned and address the coverage test issue.

Thanks for all the instructions :)

**Comment by willu47** [comment]
Hi @bernardoct. It's not a mess, so don't worry!  There's a [useful overview](https://guides.github.com/introduction/flow/index.html) of the ideal Git workflow in the Github documentation that I found really useful when I was trying to work out how to do this a while ago!

**Comment by bernardoct** [comment]
I'll read it before I my next contribution. Thanks for the help!

On Fri, Aug 28, 2015 at 9:24 AM, Will Usher <notifications@github.com>
wrote:

> Hi @bernardoct <https://github.com/bernardoct>. It's not a mess, so don't
> worry! There's a useful overview
> <https://guides.github.com/introduction/flow/index.html> of the ideal Git
> workflow in the Github documentation that I found really useful when I was
> trying to work out how to do this a while ago!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SALib/SALib/pull/73#issuecomment-135774919>.
>



-- 
Bernardo

**Comment by jdherman** [comment]
Thanks Will! The branch issue was my fault. As is the lack of unit tests for Sobol, but we'll try to fix that after we have the basics working. I wrote regression tests for the Ishigami function but nothing much more specific than that.

Right now we're just making sure (1) that the parallelization is running at the speed it should be, and (2) that the original (non-parallel) method is still intact for newcomers.

**Comment by willu47** [comment]
Great.  Looking forward to playing with the finished result.

**Comment by jdherman** [comment]
Hi @willu47, I think we have this to the point where it's ready to commit, if you have time to take a look.

It adds a nice option to parallelize the sobol indices calculation with multiprocessing while still keeping the same interface to the serial calculations. Some of the more complicated list operations are to evenly distribute the first and second order calcs across processors, since the latter take more time.

If it's alright with you I'll work on adding more tests in a separate PR -- both the existing and new functions in the sobol method need tests written.

File: SALib/analyze/sobol.py [hunk 1]
```diff
@@ -8,14 +8,18 @@
 from . import common_args
 from ..util import read_param_file
 
+from multiprocessing import Pool, cpu_count
+from functools import partial
+
+EMPTY = -1
```
Comment (willu47) [inline SALib/analyze/sobol.py]: I'm not too keen on this variable.  What does it mean?  An inline comment could help clarify what its for.  Also doesn't meet PEP standards.  Boo to ALLCAPS ;)

**Comment by willu47** [comment]
Okay, this looks good.  Thanks for running it by me!  It might be worth running some profiling on the individual methods - particularly those that are taking a long time to see if there's any possibility of optimising the existing code.

Don't forget a simple test which performs the same use of sobol, one with and one without parallel mode and compares the answer!

**Comment by bernardoct** [comment]
Hi Will,

Jon and I think it may better to commit the code as it is, as the speedup
we currently have can be considered a meaningful contribution. I've done
profiling in the code and think we are at a point in which it would be
really hard to improve its performance (at least with the time and
knowledge about parallel programming we currently have), although we 'll
keep working on it, slowly and steadily :)

I ran several tests with the Ishigami function to compare the results of
the serial and of the parallel version and the numbers look good. As Jon
said, he kindly took the responsibilities of creating the method's
individual tests.

On Sat, Aug 29, 2015 at 7:08 AM, Will Usher <notifications@github.com>
wrote:

> Okay, this looks good. Thanks for running it by me! It might be worth
> running some profiling on the individual methods - particularly those that
> are taking a long time to see if there's any possibility of optimising the
> existing code.
>
> Don't forget a simple test which performs the same use of sobol, one with
> and one without parallel mode and compares the answer!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SALib/SALib/pull/73#issuecomment-135972538>.
>



-- 
Bernardo

**Comment by willu47** [comment]
Okay, I am quite happy for Jon to merge this. Thanks again both of you for the new functionality and work on this.

**Comment by jdherman** [comment]
Ok great. I'll do the tests on a new fork, and also update the readme/examples/etc.

(By the way that EMPTY variable was removed in the refactor)

Thanks!

**Comment by bernardoct** [comment]
That EMPTY variable was me assuming that Java conventions apply to Python
out of habit. My bad.

On Sat, Aug 29, 2015 at 12:54 PM, Jon Herman <notifications@github.com>
wrote:

> Ok great. I'll do the tests on a new fork, and also update the
> readme/examples/etc.
>
> (By the way that EMPTY variable was removed in the refactor)
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SALib/SALib/pull/73#issuecomment-136009173>.
>



-- 
Bernardo

**Comment by jdherman** [comment]
This is merged into master and is now part of release v0.6.3. You can `pip install --upgrade SALib` to get it. Thanks again!

**Pull request merged**